### PR TITLE
feat : HCX-396 - HCX should generate URL for encryption and signing cert path

### PR DESCRIPTION
**Pull request by AbhiGaddi**

### CODE REVIEW

File: hcx-apis/src/main/java/org/swasth/hcx/controllers/v1/ParticipantController.java [hunk 1]
```diff
@@ -48,6 +49,15 @@ public class ParticipantController extends BaseController {
     @Value("${postgres.onboardingTable}")
     private String onboardingTable;
 
+    @Value("${certificates.aws_accesskey}")
```
Comment (shiva-rakshith) [inline hcx-apis/src/main/java/org/swasth/hcx/controllers/v1/ParticipantController.java:52]: Are these required here?

File: hcx-apis/src/main/java/org/swasth/hcx/controllers/v1/ParticipantController.java [hunk 2]
```diff
@@ -216,18 +233,21 @@ else if (notAllowedUrls.contains(requestBody.get(ENDPOINT_URL)))
...
+    }
+
+    public void validateCertificates(Map<String, Object> requestBody) throws ClientException, CertificateException, IOException {
+        if (!requestBody.containsKey(ENCRYPTION_CERT) || !(requestBody.get(ENCRYPTION_CERT) instanceof String))
             throw new ClientException(ErrorCodes.ERR_INVALID_PARTICIPANT_DETAILS, INVALID_ENCRYPTION_CERT);
-        // add encryption certificate expiry
+        else if (!requestBody.containsKey(SIGNING_CERT_PATH) || !(requestBody.get(SIGNING_CERT_PATH) instanceof String))
+            throw new ClientException(ErrorCodes.ERR_INVALID_PARTICIPANT_DETAILS, INVALID_SIGNING_CERT_PATH);
         requestBody.put(ENCRYPTION_CERT_EXPIRY, jwtUtils.getCertificateExpiry((String) requestBody.get(ENCRYPTION_CERT)));
+
```
Comment (shiva-rakshith) [inline hcx-apis/src/main/java/org/swasth/hcx/controllers/v1/ParticipantController.java:244]: Where is the signing cert path expiry field?

File: hcx-apis/src/main/resources/application.yml [hunk 3]
```diff
@@ -97,4 +97,10 @@ notification:
     allowedFilters: ${notification_subscription_allowedFilters:subscription_id, topic_code, subscription_status}
   networkPath: ${notification_networkPath:networkNotifications.yaml}
   participantPath: ${notification_participantPath:participantNotifications.yaml}
-  workflowPath: ${notification_workflowPath:workflowNotifications.yaml}
\ No newline at end of file
+  workflowPath: ${notification_workflowPath:workflowNotifications.yaml}
+
+#AWS S3 bucket
+certificates:
+  aws_accesskey : ${accesskey:}
```
Comment (shiva-rakshith) [inline hcx-apis/src/main/resources/application.yml:104]: keep the variable names in camel case

File: hcx-core/aws-client/src/main/java/org/swasth/AwsClient.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        AmazonS3 s3Client = getClient();
+    }
+
+    public AmazonS3 getClient() {
+        return AmazonS3ClientBuilder.standard()
+                .withCredentials(new AWSStaticCredentialsProvider(new BasicAWSCredentials(awsAccesskey, awsSecretKey)))
+                .withRegion(Regions.AP_SOUTH_1)
+                .build();
+    }
+    public void putObject(String folderName){
```
Comment (shiva-rakshith) [inline hcx-core/aws-client/src/main/java/org/swasth/AwsClient.java]: Give spaces between methods, maintain format.

File: hcx-apis/src/main/java/org/swasth/hcx/controllers/v1/ParticipantController.java [hunk 5]
```diff
@@ -168,6 +179,7 @@ public ResponseEntity<Object> participantDelete(@RequestHeader HttpHeaders heade
             if (!requestBody.containsKey(PARTICIPANT_CODE))
                 throw new ClientException(ErrorCodes.ERR_INVALID_PARTICIPANT_CODE, PARTICIPANT_CODE_MSG);
             String participantCode = (String) requestBody.get(PARTICIPANT_CODE);
+            awsClient.deleteMultipleObject(participantCode, bucketName);
```
Comment (shiva-rakshith) [inline hcx-apis/src/main/java/org/swasth/hcx/controllers/v1/ParticipantController.java:182]: If object is not in s3, will it throw an error.

File: hcx-apis/pom.xml [hunk 6]
```diff
@@ -100,7 +100,13 @@
 			<artifactId>commons-validator</artifactId>
 			<version>1.4.1</version>
 		</dependency>
-    </dependencies>
+		<dependency>
+			<groupId>org.swasth</groupId>
+			<artifactId>aws-client</artifactId>
```
Comment (maheshkumargangula) [inline hcx-apis/pom.xml:105]:
```suggestion
			<artifactId>cloud-storage-client</artifactId>
```

File: hcx-apis/src/main/java/org/swasth/hcx/config/AwsConfiguration.java [hunk 7]
```diff
@@ -0,0 +1,23 @@
+package org.swasth.hcx.config;
+
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.swasth.AWSClient;
+import org.swasth.ICloudService;
+
+@Configuration
+ class AwsConfiguration {
```
Comment (maheshkumargangula) [inline hcx-apis/src/main/java/org/swasth/hcx/config/AwsConfiguration.java:10]: Don't use names specific to a could service provider.

File: hcx-apis/src/main/java/org/swasth/hcx/controllers/v1/ParticipantController.java [hunk 8]
```diff
@@ -203,7 +217,7 @@ private boolean isParticipantCodeExists(String participantCode) throws Exception
         }
     }
 
-    private void validateCreateParticipant(Map<String, Object> requestBody) throws ClientException, CertificateException, IOException {
+    private void validateCreateParticipant(Map<String, Object> requestBody) throws ClientException {
```
Comment (maheshkumargangula) [inline hcx-apis/src/main/java/org/swasth/hcx/controllers/v1/ParticipantController.java:220]: Why do we have this code in controller?

File: hcx-apis/src/test/java/org/swasth/hcx/controllers/BaseSpec.java [hunk 9]
```diff
@@ -209,7 +215,34 @@ public String getParticipantCreateBody() throws JsonProcessingException {
...
         return JSONUtils.serialize(obj);
     }
-
+    public String getCertificateUrlCreateBody() throws JsonProcessingException {
+        Map<String,Object> obj = new HashMap<>();
+        obj.put("participant_name","test user");
+        obj.put("primary_mobile","9493347239");
+        obj.put("primary_email","testuser@gmail.com");
+        obj.put("certificates_type","cert_data");
```
Comment (maheshkumargangula) [inline hcx-apis/src/test/java/org/swasth/hcx/controllers/BaseSpec.java:223]: Why did you introduce this parameter in the request body? There should not be any changes to the API Spec for implementation.

**Review by maheshkumargangula — CHANGES_REQUESTED** [review]

**Review by maheshkumargangula — APPROVED** [review]

**Pull request merged**